### PR TITLE
[SPARK-38585][SQL] Simplify the code of `TreeNode.clone()`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
SPARK-28057 adds `forceCopy` arg to private `mapChildren` method in `TreeNode` to realize the `clone()` method.

After SPARK-34989, the call corresponding to `forceCopy=false` is changed to use `withNewChildren`, and `forceCopy`  always true and the private `mapChildren` only used by `clone()` method.

So this pr move the code of  private `mapChildren` method into `close()` method and remove the unreachable code branch when `forceCopy` is always true. 

### Why are the changes needed?
Code simplification.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala [hunk 1]
```diff
@@ -610,24 +610,15 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product with Tre
...
    * Returns a copy of this node where `f` has been applied to all the nodes in `children`.
+   * This method force making a copy of the nodes even if no child has been changed.
    * @param f The transform function to be applied on applicable `TreeNode` elements.
-   * @param forceCopy Whether to force making a copy of the nodes even if no child has been changed.
    */
-  private def mapChildren(
-      f: BaseType => BaseType,
-      forceCopy: Boolean): BaseType = {
-    var changed = false
+  private def mapChildrenWithForceCopy(f: BaseType => BaseType): BaseType = {
```
Comment (LuciferYang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:616]: Is there a more suitable method name?

Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:616]: Should we move this method into `clone()` method? It is now only used by `TreeNode.clone`

**Review by srowen — APPROVED** [review]
The change checks out, after understanding the possible simplifications.

**Comment by sigmod** [comment]
cc @dbaliafroozeh @hvanhovell

**Comment by sigmod** [comment]
cc @maryannxue

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by LuciferYang** [comment]
thanks all

**Pull request closed without merging**